R/runCalibration.R

```diff
@@ -1291,50 +1273,114 @@ runCalibrationOptim <- function(path,
+    stepSizeParamsSelect <- mutate(stepSizeParamsSelect, stepSize = 0)
+
+    stepSizeParams %>%
+      anti_join(stepSizeParamsSelect, by = c("region", "loc", "typ", "inc")) %>%
+      rbind(stepSizeParamsSelect)
+  } else {
+    message("For at least one subset: ",
+            "The armijo condition is not satisfied in the predefined number of iterations of the step size algorithm",
+            "and no minimum has been found.",
+            "Since the outer objective has been decreasing throughout the iterations, the last step size is used.")
   }
-  stepSizeParamsSelect <- mutate(stepSizeParamsSelect, stepSize = ifelse(
-    .data$minStepSize != 0,
-    .data$minStepSize,
-    .data$stepSize
-  ))
+}
 
-  stepSizeParams %>%
-    anti_join(stepSizeParamsSelect, by = c("region", "loc", "typ", "inc", "ttot")) %>%
-    rbind(stepSizeParamsSelect)
+#' Compute the sum of the squares for a calibration target
+#'
+#' @param target data frame with calibration target data
+#' @param zeroFlow logical, whether data should be filtered for the zero flow
+#'
+#' @importFrom dplyr across all_of group_by summarise
+#'
+.computeSumSqTarget <- function(target, zeroFlow = FALSE) {
+  if (isTRUE(zeroFlow)) {
+    for (column in intersect(c("bsr", "hsr"), colnames(target)))
+    target <- filter(target, .data[[column]] == "0")
+  }
+
+  target %>%
+    group_by(across(all_of(c("region", "loc", "typ", "inc")))) %>%
+    summarise(target = .sumSquare(.data$target), .groups = "drop")
+}
+
+#' Check for the stopping criterion of the iteration
+#'
+#' Returns \code{TRUE} if and only if both the absolute and the relative stopping criterion
+#' are satisfied for all subsets.
+#'
+#' @param outerObjective data frame containing the value of the outer objective function.
+#' @param p_calibTarget list of data frames with historic flows
+#' @param threshold list of numeric, absolute and relative threshold of the stopping criterion
+#' @param zeroFlow logical, whether renovation flow data should be filtered for zero flows
+#'
+.checkStoppingCriterion <- function(outerObjective, p_calibTarget, threshold, zeroFlow = FALSE) {
+  sumSqTarget <- do.call(rbind, lapply(names(p_calibTarget), function(var) {
+    .computeSumSqTarget(p_calibTarget[[var]], zeroFlow && grepl("renovation", var))
+  })) %>%
+    group_by(across(all_of(c("region", "loc", "typ", "inc")))) %>%
+    summarise(target = sum(.data$target), .groups = "drop")
+
+  outerObjective %>%
+    select(-"fPrev") %>%
+    left_join(sumSqTarget, by = c("region", "loc", "typ", "inc")) %>%
+    mutate(absLevel = sqrt(.data$f),
+           relLevel = sqrt(.data$f) / sqrt(.data$target),
+           .keep = "unused") %>%
+    mutate(absCriterion = .data$absLevel <= threshold[["abs"]],
+           relCriterion = .data$relLevel <= threshold[["rel"]],
+           .keep = "unused") %>%
+    pivot_longer(cols = c("absCriterion", "relCriterion"), names_to = "criterion") %>%
+    pull("value") %>%
+    all()
```

Comment:
I think `all` can handle data frames with multiple columns.
```suggestion
           .keep = "none") %>%
    all()
```